### PR TITLE
AIP-69: Add Executor to Edge Provider

### PR DESCRIPTION
**Pull request by jscheffl**
This PR is a split of https://github.com/apache/airflow/pull/41729 and builds on-top of https://github.com/apache/airflow/pull/42046 and #42047 to make review easier.

It only contains the new Edge Executor functions needed of the edge provider package. The first commit is from the other base PRs and if this is merged will go-away on rebase. If you want to see Executor implementation only, please take a look to the second commit only,

### CODE REVIEW

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [inline airflow/providers/edge/executors/edge_executor.py]: Shouldn't this do something to block and flush running tasks?

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [inline airflow/providers/edge/executors/edge_executor.py]: Most executors use an in-memory queue of some kind to ingest tasks and track them for execution. Using the DB like this is certainly more durable (though you only call commit in the sync method) but we're going to get a lot of write load to the DB using the database like this (if you have an airflow environment with very high Task TPS). Have you done any benchmarking to see the impact on the DB?

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [inline airflow/providers/edge/executors/edge_executor.py:1]: What PR (maybe already merged?) contains the code where tasks are actually sent over API to the workers?

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [inline docs/apache-airflow-providers-edge/edge_executor.rst:65]: I believe I saw in another PR you added this executor to the holy list of "core" executors. So folks should be able to use the short name instead of the module path right?

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [inline docs/apache-airflow-providers-edge/edge_executor.rst]:
```suggestion
  - ``api_enabled``: Must be set to true. It is disabled intentionally to not expose
```

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [inline docs/apache-airflow-providers-edge/edge_executor.rst]: Is this what was meant here?:
```suggestion
is included in the provider package and install it on the webserver and use the
```

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [inline docs/apache-airflow-providers-edge/edge_executor.rst:99]: Should this get some kind of heading or bolding?

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py:1]: As the scheduler runs in the "backend" and is not exposing an API, the tasks are fetched/polled via a REST API hosted in webserver (with the DB as sync point between).
The API is exposed in https://github.com/apache/airflow/pull/42049 - this is leveraging the AIP-44 internal API and adds the method `EdgeJob.reserve_task` on the RPC for this in https://github.com/apache/airflow/pull/42049/files#diff-72e5d2c8bdcd2dc42c0cceb6f5f81b032a0a0a6fec52917b8d95c0daac9e45a8R63

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (jscheffl) [reply docs/apache-airflow-providers-edge/edge_executor.rst:65]: Yes, for Core this is in https://github.com/apache/airflow/pull/41730

For Breeze it is in https://github.com/apache/airflow/pull/41731

(Separated as we have the rule to split provider and core PRs to cherry-pick)

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: I was also thinking about this. From the position of (potentially) multiple scheduler instances and the structure that open jobs are in the DB table... actually all schedulers contribute to the queue of jobs. No scheduler "owns" a job but just distributes them.

I understand that a graceful stop is needed for a Local Executor. But in the case of Edge the jobs are running remote. Stopping the scheduler will just end adding new jobs, existing (queued) jobs can continue also after stopping the scheduler. Is it a required interface to wait for all tasks? What if another scheduler is adding more jobs... then the scheduler going to stop would potentially never terminate. Or we need to tag which were the tasks created by this scheduler instance to know when which instance is eligible to "end".

For sake of simplicity I thought we can just stop.

If you think I need to drain all jobs started, then I would need to add a draining of (remote running) jobs. In my tests stopping the scheduler did not harm the completion (async). Let me know. (I am not an expert on the scheduler interface...
(Following the docs in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html the implementation is optional though)

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: I was thinking about this all-day. Do you know a "simple" method to count the number of read/write operations in the DB? (Not transactions but statements?) Then I'd offer to make a measurement.

I would assume that (1) this is a MVP implementation and is only for a small functional test. In the AIP I was offering to add scalability later.

From gut feeling w/o measuring I assume the write transactions are trivial and IOPS not much compared to other queries made by the scheduler. Also comparing to Celery - Celery also has a result backend table where each job adds a tuple into the "job"(?) table - I assume this is very comparable. I feel the overhead on DB is quire small.

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: > Or we need to tag which were the tasks created by this scheduler instance to know when which instance is eligible to "end".

Tasks are already owned by a specific scheduler, you will see dag runs and task instances all have the job_id of the scheduler job runner which started them. This is how failing stuck tasks and task adoption all works. If the job runner that ran the tasks dies/is killed (i.e. it doesn't heartbeat for some period of time) the TIs will be adopted by another scheduler job runner (if your executor has task adoption implemented, which this one doesn't) or be reset if they are not adopted.

So not mandatory but you will get funny stuff happening if you don't drain your tasks or at least support adoption. This executor is doing neither. But things will also work if your tasks are short enough to not hit that window.

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [reply docs/apache-airflow-providers-edge/edge_executor.rst:65]: So in that case the value does not need to be the module path, it can just be "EdgeExecutor".

Btw, how did we decide to add another executor to that holy list? I thought we were no longer doing so?

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: > I was thinking about this all-day. Do you know a "simple" method to count the number of read/write operations in the DB? (Not transactions but statements?) Then I'd offer to make a measurement.

I'm not sure, but either way I think the onus is on the author to test their code. Much time was spent during the hybrid executor project to get benchmarking of some kind (that wasn't available off the shelf).

> I would assume that (1) this is a MVP implementation and is only for a small functional test. In the AIP I was offering to add scalability later.

That's fair, I just worry once these things get in the door the tech debt isn't cleaned up. But you are very diligent and present so I don't worry much in this case if you plan to come back and fix later :smiley: 

> From gut feeling w/o measuring I assume the write transactions are trivial and IOPS not much compared to other queries made by the scheduler. Also comparing to Celery - Celery also has a result backend table where each job adds a tuple into the "job"(?) table - I assume this is very comparable. I feel the overhead on DB is quire small.

Perhaps, but do note that this all adds up now, if you run celery along side edge executors in a multiple executor configuration (and we continue to allow other execs to use the DB) we could find ourselves in a situation with a _lot_ of traffic to the DB.

At the end of the day, I'm no DB expert, so perhaps someone can weigh in. It's just making my spidey-senses tingle when I saw this.

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (jscheffl) [reply docs/apache-airflow-providers-edge/edge_executor.rst:65]: Oh, I did not know that the list of executors is "holy" - was there a decision that only specific "premium" Executors shall be added? Maybe I am too new or missed this. So if I "should not" then this would just be affecting #41730. I "naively thought" having the full module path here until the other PR is merged, then with a future Airflow version the short name could be used. But I would not insist on a short name. You/the community is able to have a word on this and convince me of course.

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Yes. Thanks for the remarks. I did not intend to make the implementation final or "hide" tech debt. It is marked as MVP to provide a first function and the aim is to incrementally improve. This means in functionality, reliability and performance.

My intend after MVP is to use it and step by step improve open items based on the documented backlog - by myself, my team - and if somebody else is using also happy to receive contributions from the community.

Just added some notes to the RST to make clear that performance tuning is something that will/must be added post MVP. So no load tests have been made until now.

And also I acknowledge that there was a lot of performance consideration in current Hybrid Executor implementation. As well as in Non-Hybrid K8s Executor. Was also not aware (before the summit) about the async-complexity - But also in case of K8s you have a lot of remote calls on the API to raise and follow started workload. And assuming a many users expect scalability to scale on K8s/Celery as of today.

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (o-nikolas) [reply docs/apache-airflow-providers-edge/edge_executor.rst:65]: Yeah, that list is really only for executors that airflow vends, they're sometimes called "core" executors. Local, Sequential, Debug are all examples. Celery and Kubernetes now live in provider packages, but are grandfathered into being in this holy list because they existed before executors lived in provider packages. Since your executor exists in a provider package, my vote would be to _not_ add it to the holy list in the executor loader and then if that's the case, then users would then need the full module path like you have in the docs now.

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: In my "Integration Test" DAG (that is alread ymerged from previous PRs and sitting in `airflow/providers/edge/example_dags/integration_test.py`) I added a long running task (15min) to ensure a heartbeat is tested. That all is running fine. Also thanks to `tenancity` added to communication layer a disconnect of ~5min is survived and task still is in success.

But your comment reminds me that I wanted to make two stability tests which I forgot:
1. Stop and restart the scheduler after some time and check that all tasks will run to finish
2. kill -9 the worker and task run process and ensure that after missing heartbeat the task is marked with error and leftovers are cleaned in the edge_job table.

I'll double check this... tomorrow.

File: docs/apache-airflow-providers-edge/edge_executor.rst
Comment (jscheffl) [reply docs/apache-airflow-providers-edge/edge_executor.rst:65]: Okay, then I'll adjust #41730 and revert 99% of the changes there. Then the docs here will stay as it is fr the time being. (Until this one is stable, performant, cool and getting grandfathered :-D)

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Hi @o-nikolas ... took a moment but now I was able to make the stability checks, this is the outcome:

Tests were based in the "Integration Test" DAG which contains all tested functions and operators which is added to provider package - will be base of system test in future if I get to this point. As you asked for heartbeat - there is one "long running" tasks that sleeps 1 minute and sends some logs for 15 minutes runtime - so a task that is hitting the heartbeat timeout. Added this before to ensure heartbeating is working and long running tasks can be made on the edge.

- [x] Positive case - let it run - SUCCESS - long running task reported OK
- [x] Kill -9 the "task run" process - FAILED as expected - task was reported as failed (immediately). Logs are expected like:
      `INFO - Task exited with return code -9For more information, see https://airflow.apache.org/docs/apache-airflow/stable/troubleshooting.html#LocalTaskJob-killed`
- [x] Kill -9 edge worker, leave forked long task process running - SUCCESS (because task completes also w/o worker and updates the state itself via task runner API)
      Side effect is that logs are not uploaded as this is done by the worker which is dead.
      After the restart of Edge Worker, the DAG picked the last task and ended with success.
      Future improvement will be to upload leftover logs after restart of edge worker
- [x] Kill -9 edge worker and task runner (like if a container or server dies - FAILED as expected
      After 300(+) seconds - the task was detected as zombie and was set to failed by the scheduler.
      No need to have something special implemented on the EdgeExecutor or Edge Worker (which was dead anyway)
- [x] Cut network connectivity (Actually stop webserver/API) - FAILED - as expected
      Edge worker as well as task runner tried via @tenacity.retry() to retry 10 times which took ~9minutes. Then gave-up and terminated.
      If the webserver/connection recovers within the retry attempts the task will continue and all problems are masked.
      In my case I did not restart webserver. So Edge Worker and task runner terminated/gave-up calling.
      In the scheduler the running task was reported as zombie as exected.

As all Executors work on the common DB table and API calls go indirect via DB... in the Executor there is no real ownership needed in my view. Also no adoption. Based on the DB table Executors just report back the "running" tasks for the planning of PARALLELISM.

Would this be OK for you? Or do you see an task adoption is needed?

Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: Love the testing! Thanks for doing that!
You didn't see task adoption because you didn't run that test case, adoption happens only after the scheduler dies and a new scheduler is started. The adoption code will detect that the scheduler job that used to own those tasks is now gone and another scheduler process job will try to adopt it.

As for it being needed, that's up to you I suppose. It's not a mandatory part of the base executor interface, but I think a valuable one, personally.

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Ah, thanks, now got it... and made 3 more tests:

- [x] While the long running task was running (2 of 15mins) I stopped the scheduler. Task completed w/ success.
      Restarted Scheduler after the task completion and the rest of the tasks of the DAG were scheduled. All OK.
      Task execution is (and should not) depend on the scheduler
- [x] Started three schedulers which were competing around scheduling. The test DAG was scheduled and executed in success. All OK
- [x] Stopped Scheduler and restarted while the long running task was (happily) running. Restarted Scheduler killed the (happily) running tasks and marked as failed. NOT EXPECTED.

Will check for the third case. Probably this is where task adoption kicks-in :-D

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Fixed adoption problem with https://github.com/apache/airflow/pull/42048/commits/6011d7ec7d2cd1506ff23169ea72d9a8313aa3e7

Anything that you want to have changed or OK to approve/merge in the current state?

Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: > Will check for the third case. Probably this is where task adoption kicks-in :-D

Yupp :) 

> Fixed adoption problem with https://github.com/apache/airflow/commit/6011d7ec7d2cd1506ff23169ea72d9a8313aa3e7

> Anything that you want to have changed or OK to approve/merge in the current state?

Looking now!

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [inline airflow/providers/edge/executors/edge_executor.py]: Wait, I'm only now realizing this. So every executor (in a multiple scheduler/HA environment) is reading all tasks across the airflow cluster? This is another unusual sideeffect of how you've implemented this executor. Usually executors maintain a queue of tasks that they've received from _only_ their scheduler, but in this case your executor implementation is going to be stepping over the toes of every other instance of the edge executor in an HA environment.

Is this intended?

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: For the MVP it is intended.

I would still refer to the fact that this is MVP. Main intent is to have something running and then to optimize.
I would not say the schedulers are stepping on each-others toe, they just share one DB table for all open jobs. Jobs are purged from the table a few minutes after completion. So it is all very simple short running transactions and all should be in cache of Postgres. The shared table contains N*PARALLELSIM open records for N schedulers.

For AIP for MVP I documented that max 100 workers are the planned start. Later improvement:
image

But as there might be doubt... just created a small DAG with a dynamic mapped PythonOperator, 1 print() statement. Mapped to 500 tasks. Executed on my laptop with webserver, 1 scheduler, 4 worker (each concurrency of 8) and the 500 tasks were completed in 11:48min. Whereas almost all time was used by python processes forked off workers to start/init. Second biggest CPU consumer was webserver (~1 core) for the API backend. Scheduler as well as Postgres DB each consumed ~3-5% of one of the cores.
Have not more hardware at my hands to distribute more load by workers but scheduler in this small test was not showing any performance challenge.

File: airflow/providers/edge/executors/edge_executor.py
Comment (ashb) [reply airflow/providers/edge/executors/edge_executor.py]: Yeah, scheduler stopping shouldn't terminate the tasks -- Celery and Kube executor don't do that for instance.

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Then... I assume we are okay here.

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: @o-nikolas Had a longer thought on this and as I am still seeking for other reviewers... two arguments towards "every executor is reading all tasks across the cluster":

1. During scheduling (which is the rather performance critical activity) just DB entries are added. SQLalchemy will batch them up and flush them in the scheduling loop. This should be pretty efficient.
2. All tasks are _only_ read from the table in the `sync()` call in order to clean up the internal state and the table itself. `sync()` is called periodically and as it is a cleanup it is mainly to remove old tasks and check for completion (to schedule more). Besides binding tasks to a specific scheduler instance we could also reduce the frequency of cleanup to say for example only clean-up every N `sync()` calls. I saw this being executed ~every second. Therefore I took a look to (current) timing on my laptop and came to:
  - If 1 task is in the edge_job table then the `sync()` needs 1-2ms per call (which seems to happen roughly every second)
  - If I added 1000 artificial tasks the time for `sync()` increased to 25-30ms.
  - Therefore I took a look to the code, optimized it one round in https://github.com/apache/airflow/pull/42048/commits/c2672ed2ac356de6f925f1be9474d4fba678ae48 which filters the rows for relevance in a first optimization to the relevant states for the check in `sync()`. With this if 1000 tasks are queued the time drops down to 2-3ms and will only increase with the amount of running or tasks to clean.
3. If you compare the "additional DB load" to Celery - also in Celery per task tuples are added and updated in the result table. Did not count the number of transactions but it is also one row per task in a global table `celery_taskmeta`. With Edge Executor we are in a similar range.

Does this improve your opinion about DB overhead and performance? Otherwise please let me know if you would request me to filter rows per scheduler down already in MVP. (We currently use 3 scheduler instances, in my view the overhead is quite minimal for the sync).

Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: Two thoughts:

1) You've taken great care in this case to make the DB interaction fairly optimized. Thanks for that! But a big concern is that we're setting a _precedent_ here, for folks to use the DB as the queue for tasks in their executors. With multiple executors running concurrently and others perhaps not taking as much care as you have, could lead to poorer outcomes. I think as a community and piece of software, we are agreed on removing as many of our direct dependencies on the metadata DB and strongly encouraging contributors to not write code that expects to have direct access to it like this. I think this executor as implemented works against that goal.

2) Not only the DB impact, but the logical impact: up until this point each executor/scheduler pair owned a specific set of tasks. And there are assumptions in Airflow logic based on that. This implementation means that all executors are sharing tasks because they are communicating over the DB. I can't sign off on the implications of this honestly, only that this is not the norm for executors and I'm not sure what the downstream effects of that could be.

Happy to have this PR merged, I'd just be curious to see if others have the same concerns. If not, then feel free to get approvals and merge! :smiley:

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: UPDATE after initial post matching your concerns:

Regarding (1): Forgot to mention: The EdgeExecutor has the target (together with AIP-72) to un-bundle the DB from the task execution. Here in Airflow 2.10 only partial targets can be achieved. For the execution scheduling itself the Executor needs to have access to DB. I believe the move to Edge Executor is rather paying into the targets not adding new dependencies. And it should be only the Scheduler in _this_ PR be the only component with DB Access (Except current state in Airflow 2.10 the internal API bridges this as well until AIP-72 is on main)

Regarding (2): We could optimize the global vs. local access per Scheduler. Made a load test as below and did not see a real limitation. Making the queries to DB based on the Scheduler instance first-hand would make DB calls rather more expensive than faster in my checks. Even with 3 Scheduler under load the results are very good and performance optimizations are possible but limiting factors are not the Scheduler.

Regarding Load:

As of DB impact concerns I made a load test yesterday in our Azure AKS environment and used K8s PODs for workers inside the cluster. Based on the current code my results were:

Setup:
- 3 Scheduler (PODs)
- 30 Edge Worker, each concurrency=6
- Postgres DB, 8 Cores, 32GB RAM (Azure Flexible Server)
- 1 DAG with 1 Mapped Task, 1000 mappings. Second DAG which calls the DAG i a batch

Results:
- Constant execution needed 2:30min per 1000 tasks/DAG (120 runs, all 120 000 tasks in success, no flakiness)
- DB Load was up ~200 IOPS and +20% CPU over the base load
- 3 Scheduler PODs each needed ~35% of a single core
- Load was very constant over multiple hours, would have achieved 550 000 tasks in a day

Main limiting factor and driver for DB load seems to be rather the Edge Worker making a lot of API requests --> DB via JSON RPC. Needed to scale Webserver and PGBouncer following as of the DB queries mainly from Webserver.

I would not have any fears in regards of DB load seen from monitoring that was emitted by Scheduler itself. Rather the "chattiness" from the AIP-44 internal API will need to be replaced with upcoming AIP-72 if there should be any scaling promise. But of course the Scheduler also can be optimized... post MVP.

File: providers/src/airflow/providers/edge/executors/edge_executor.py
Comment (ashb) [inline providers/src/airflow/providers/edge/executors/edge_executor.py]: This raising feels like it would end up blowing up the scheduler?

**Review by ashb — APPROVED** [review]
Pre-emptively approving, but please either address https://github.com/apache/airflow/pull/42048/files#r1805240595 or tell me why it's not a problem.

File: providers/src/airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/executors/edge_executor.py]: I double checked and in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/index.html#optional-interface-methods-to-implement it is listed as optional implementation. BaseOperator also has the same code, so same exception is raised if not implemented.

The function is called in https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job_runner.py#L1802 where explicitly this is catched and a debug is logged.

But as this is use-less as it is the same like in base-class, let me remove this and put it up to the open item list to improve before a v1 release of the provider.

File: airflow/providers/edge/executors/edge_executor.py
Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: @o-nikolas I added explicit notes into the RST of the executor that a performance assessment ensuring no side effect/impact on other executors before v1.0.0 release.. OK then to merge?

**Comment by jscheffl** [comment]
> Pre-emptively approving, but please either address https://github.com/apache/airflow/pull/42048/files#r1805240595 or tell me why it's not a problem.

Can confirm both: Addressed it and it is not a problem!

File: airflow/providers/edge/executors/edge_executor.py
Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: Sure, no worries! Regarding my points 1 and 2 above, it's not entirely about performance. But I'm glad you got some other points of view on this one (assuming others reviewed this and silently agreed it's okay).

Comment (o-nikolas) [reply airflow/providers/edge/executors/edge_executor.py]: Also, I'm working on some multi-team stuff lately and realized that the way this executor shares tasks between different instances will likely not work for that kind of setup. In multi-team executors should be siloed to a specific team, and should not have access to the tasks from other teams. This executor will definitely require updates to support being used in a multi-team capacity.

Comment (jscheffl) [reply airflow/providers/edge/executors/edge_executor.py]: Thanks for the heads-up. Yes I am sure that we can not this as a limitation for multi-team. As it is a MVP, not all use cases are supported. If it matures and this will be a requirement that is needed I am fine to extend this. Also for large scaling maybe alternative queues to DB can be integrated. Not limited to stay as it is and I am committed to improve!

**Pull request merged**
